data/secscan_model/secscan_v4_model.py

```diff
@@ -0,0 +1,157 @@
+        except ManifestSecurityStatus.DoesNotExist:
+            return SecurityInformationLookupResult.with_status(ScanLookupStatus.NOT_YET_INDEXED)
+
+        if status.index_status == IndexStatus.FAILED:
+            return SecurityInformationLookupResult.with_status(ScanLookupStatus.FAILED_TO_INDEX)
+
+        if status.index_status == IndexStatus.UNSUPPORTED:
+            return SecurityInformationLookupResult.with_status(
+                ScanLookupStatus.UNSUPPORTED_FOR_INDEXING
+            )
+
+        if status.index_status in [
+            IndexStatus.TIMED_OUT,
+            IndexStatus.WAITING,
+            IndexStatus.IN_PROGRESS,
+        ]:
+            return SecurityInformationLookupResult.with_status(ScanLookupStatus.NOT_YET_INDEXED)
+
+        assert status.index_status == IndexStatus.COMPLETED
+
+        # TODO(alecmerdler): Fetch data from Clair API
+        return SecurityInformationLookupResult.with_status(ScanLookupStatus.NOT_YET_INDEXED)
+
+    def perform_indexing(self, start_token=None):
+        whitelisted_namespaces = self.app.config.get("SECURITY_SCANNER_V4_NAMESPACE_WHITELIST", [])
+
+        def eligible_manifests(base_query):
+            if len(whitelisted_namespaces) == 0:
+                return base_query
+
+            return (
+                base_query.join(Repository)
+                .join(User)
+                .where(User.username << whitelisted_namespaces)
+            )
+
+        min_id = (
+            start_token
+            if start_token is not None
+            else Manifest.select(fn.Min(Manifest.id)).scalar()
+        )
+        max_id = Manifest.select(fn.Max(Manifest.id)).scalar()
+
+        if max_id is None or min_id is None or min_id > max_id:
+            return None
+
+        # FIXME(alecmerdler): Filter out any `Manifests` that are still being uploaded
+        def not_indexed_query():
+            return (
+                eligible_manifests(Manifest.select())
+                .switch(Manifest)
+                .join(ManifestSecurityStatus, JOIN.LEFT_OUTER)
+                .where(ManifestSecurityStatus.id >> None)
+            )
+
+        def index_error_query():
+            return (
+                eligible_manifests(Manifest.select())
+                .switch(Manifest)
+                .join(ManifestSecurityStatus)
+                .where(ManifestSecurityStatus.index_status == IndexStatus.FAILED)
+            )
+
+        def needs_reindexing_query():
+            # TODO(alecmerdler): Retrieve `/state` from security scanner API
+            indexer_hash = ""
+            return (
+                eligible_manifests(Manifest.select())
+                .switch(Manifest)
+                .join(ManifestSecurityStatus)
+                .where(ManifestSecurityStatus.indexer_hash != indexer_hash)
+            )
+
+        # 4^log10(total) gives us a scalable batch size into the billions.
+        batch_size = int(4 ** log10(max(10, max_id - min_id)))
```

Comment:
We should likely move this into a method, that takes in a query modifier, and then just yield from all three or somesuch

Reply:
Switched to use [`itertools.chain()`](https://docs.python.org/3/library/itertools.html#itertools.chain) instead.